[LCP] Implement LargestContentfulPaint for text

smfr opened the pull request:
#### e9cb09c7dcced3bcc75c2057d1de73facc34d320
<pre>
[LCP] Implement LargestContentfulPaint for text
<a href="https://bugs.webkit.org/show_bug.cgi?id=299832">https://bugs.webkit.org/show_bug.cgi?id=299832</a>
<a href="https://rdar.apple.com/161609007">rdar://161609007</a>

Reviewed by Tim Nguyen.

Start calling `Document::didPaintText()` from `TextBoxPainter::paint()` to track text paints
for LCP. Pass the `formattingContextRoot` since we have a rect in its coordinate system already.

In `LargestContentfulPaintData::didPaintText()` we have to find the containing block
per spec[1], after which we can get to the element to check if it&apos;s in the set of
textful elements that we&apos;ve already seen. If no, we add to `m_paintedTextRecords`, unioning
the rect with any existing rect for this element (which is effectively computing a bounds
for all the painted text nodes, per spec).

`LargestContentfulPaintData::takePendingEntry()` then needs to consider `m_paintedTextRecords`
when generating the potential LCP entry.

Add test results for the newly passing tests.

[1] <a href="https://w3c.github.io/paint-timing/#sec-modifications-dom">https://w3c.github.io/paint-timing/#sec-modifications-dom</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/element-only-when-fully-active-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/first-letter-background-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/first-paint-equals-lcp-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-full-viewport-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/larger-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-css-generated-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-mathml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-random-namespace-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/text-with-display-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/toJSON-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-with-shadow-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-with-text-stroke-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/update-on-style-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/web-font-styled-text-resize-block-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/web-font-styled-text-resize-swap-smaller-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-deadline-shortened-by-rendering-update.html: The test needs to wait another
event loop turn for the LCP-related update triggered by the `description()` rendering.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didPaintText const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaint.cpp:
(WebCore::LargestContentfulPaint::element const):
* Source/WebCore/page/LargestContentfulPaint.h: Use a WeakPtr for the element to avoid document leaks.
This is OK, since if the element is not connected, this function returns null per spec.
* Source/WebCore/page/LargestContentfulPaintData.cpp:
(WebCore::LargestContentfulPaintData::potentiallyAddLargestContentfulPaintEntry):
(WebCore::LargestContentfulPaintData::takePendingEntry):
(WebCore::LargestContentfulPaintData::didPaintText):
* Source/WebCore/page/LargestContentfulPaintData.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::mapLocalToContainer): Early return, just like `RenderBox::mapLocalToContainer()` does.

Canonical link: <a href="https://commits.webkit.org/300790@main">https://commits.webkit.org/300790@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fabddf2d3d769d6da28850b4a3d600d7db58301e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123895 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43610 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130682 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/884d4ebc-5a20-4c3c-b541-9569bd13273a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125772 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44334 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52204 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3be28f2e-040e-41e8-bb9c-b28cf803bf3b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126849 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/35323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2f892ce0-7a27-46c8-ad0d-50ba22e1c691) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/28995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74158 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/105051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29218 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133377 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50848 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51222 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107054 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/102533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26063 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/47875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26126 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50701 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56465 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/50175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/LargestContentfulPaintData.cpp:
```diff
@@ -313,8 +324,43 @@ void LargestContentfulPaintData::didPaintImage(Element& element, CachedImage* im
 {
+    auto& renderBlockFlow = const_cast<RenderBlockFlow&>(formattingContextRoot);
+    // https://w3c.github.io/paint-timing/#sec-modifications-dom says to get the containing block.
+    CheckedPtr<RenderBlock> containingBlock = &renderBlockFlow;
+    if (containingBlock->isAnonymous()) {
+        CheckedPtr ancestor = containingBlock->firstNonAnonymousAncestor();
+        if (CheckedPtr ancestorBlock = dynamicDowncast<RenderBlock>(ancestor.get()))
+            containingBlock = ancestorBlock;
+        else
+            containingBlock = containingBlock->containingBlock();
+    }
```
Does `->container()` return the right thing here without the extra hoops for anonymous renderers?

smfr replied:
No, and this is tested by some of the WPT.

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -1307,6 +1307,24 @@ imported/w3c/web-platform-tests/largest-contentful-paint/invisible-images.html [
 imported/w3c/web-platform-tests/largest-contentful-paint/larger-image.html [ Pass ]
 imported/w3c/web-platform-tests/largest-contentful-paint/supported-lcp-type.html [ Pass ]
 imported/w3c/web-platform-tests/largest-contentful-paint/video-play-after-poster.html [ Pass ]
+imported/w3c/web-platform-tests/largest-contentful-paint/element-only-when-fully-active.html [ Pass ]
```
ideally we'd flip the gardening to listing fails (could be a separate commit)

smfr committed:
[LCP] Implement LargestContentfulPaint for text
https://bugs.webkit.org/show_bug.cgi?id=299832
rdar://161609007

Reviewed by Tim Nguyen.

Start calling `Document::didPaintText()` from `TextBoxPainter::paint()` to track text paints
for LCP. Pass the `formattingContextRoot` since we have a rect in its coordinate system already.

In `LargestContentfulPaintData::didPaintText()` we have to find the containing block
per spec[1], after which we can get to the element to check if it's in the set of
textful elements that we've already seen. If no, we add to `m_paintedTextRecords`, unioning
the rect with any existing rect for this element (which is effectively computing a bounds
for all the painted text nodes, per spec).

`LargestContentfulPaintData::takePendingEntry()` then needs to consider `m_paintedTextRecords`
when generating the potential LCP entry.

Add test results for the newly passing tests.

[1] https://w3c.github.io/paint-timing/#sec-modifications-dom

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/element-only-when-fully-active-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/first-letter-background-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/first-paint-equals-lcp-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/image-full-viewport-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/larger-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-css-generated-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-mathml-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-random-namespace-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/observe-text-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/text-with-display-style-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/toJSON-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-with-shadow-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/transparent-text-with-text-stroke-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/update-on-style-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/web-font-styled-text-resize-block-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/largest-contentful-paint/web-font-styled-text-resize-swap-smaller-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/requestidlecallback/requestidlecallback-deadline-shortened-by-rendering-update.html: The test needs to wait another
event loop turn for the LCP-related update triggered by the `description()` rendering.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::didPaintText const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LargestContentfulPaint.cpp:
(WebCore::LargestContentfulPaint::element const):
* Source/WebCore/page/LargestContentfulPaint.h: Use a WeakPtr for the element to avoid document leaks.
This is OK, since if the element is not connected, this function returns null per spec.
* Source/WebCore/page/LargestContentfulPaintData.cpp:
(WebCore::LargestContentfulPaintData::potentiallyAddLargestContentfulPaintEntry):
(WebCore::LargestContentfulPaintData::takePendingEntry):
(WebCore::LargestContentfulPaintData::didPaintText):
* Source/WebCore/page/LargestContentfulPaintData.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::mapLocalToContainer): Early return, just like `RenderBox::mapLocalToContainer()` does.

Canonical link: https://commits.webkit.org/300790@main
